### PR TITLE
Add Laminas Integration

### PR DESCRIPTION
**Pull request by PROFeNoM**
### Description

Adds a Laminas integration. The goal was to capture the overall order of events and their corresponding listeners to understand the flow of execution better.

Since Laminas follows semantic versioning and does not carry a single framework version, the versions tested in this integration correspond to the [skeleton application ones](https://github.com/laminas/laminas-mvc-skeleton).

### Readiness checklist
- [ ] (only for Members) Changelog has been added to the release document.
- [X] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: src/Integrations/Integrations/Laminas/LaminasIntegration.php [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+            'attach',
+            null,
+            function ($This, $score, $args) use ($MVCEvents) {
+                if (!in_array($args[0], $MVCEvents)) {
+                    return;
+                }
+
+                $listener = $args[1];
+                $className = get_class($listener[0]);
+                $methodName = $listener[1];
```
Comment (bwoebi) [inline src/Integrations/Integrations/Laminas/LaminasIntegration.php:86]: $listener is a callable. Probably all uses you've checked are `[$class, $method]`? But I think this should at least check whether it's actually an array.

(I suppose DDTrace\install_hook should be extended to generally accept any callable, also these written as `["class", "method"]` ...)

File: src/Integrations/Integrations/Laminas/LaminasIntegration.php [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        ];
+
+        $integration = $this;
+
+        hook_method(
+            'Laminas\EventManager\EventManagerInterface',
+            'attach',
+            null,
+            function ($This, $score, $args) use ($MVCEvents) {
+                if (!in_array($args[0], $MVCEvents)) {
```
Comment (pierotibou) [inline src/Integrations/Integrations/Laminas/LaminasIntegration.php:80]: Maybe add a pointer to the code we're instrumenting, that could help understand what is contained in $arg

File: src/Integrations/Integrations/IntegrationsLoader.php [hunk 3]
```diff
@@ -75,6 +76,8 @@ public function __construct(array $integrations)
             '\DDTrace\Integrations\Laravel\LaravelIntegration';
         $this->integrations[MysqliIntegration::NAME] =
             '\DDTrace\Integrations\Mysqli\MysqliIntegration';
+        $this->integrations[LaminasIntegration::NAME] =
+            '\DDTrace\Integrations\Laminas\LaminasIntegration';
```
[the remark this answers is not in the document]
Comment (pierotibou) [reply src/Integrations/Integrations/IntegrationsLoader.php:80]: Can we explicit that with a comment in the code somewhere?

File: ext/integrations/integrations.c [hunk 4]
```diff
@@ -163,6 +163,11 @@ void ddtrace_integrations_minit(void) {
     DD_SET_UP_DEFERRED_LOADING_BY_METHOD(DDTRACE_INTEGRATION_ELOQUENT, "Illuminate\\Database\\Eloquent\\Model", "destroy",
                                          "DDTrace\\Integrations\\Eloquent\\EloquentIntegration");
 
+    DD_SET_UP_DEFERRED_LOADING_BY_METHOD(DDTRACE_INTEGRATION_LAMINAS, "Laminas\\Mvc\\Application", "init",
+                                         "DDTrace\\Integrations\\Laminas\\LaminasIntegration");
```
Comment (pierotibou) [inline ext/integrations/integrations.c:167]: No chance to load the integration twice as we react to 2 methods?

File: src/Integrations/Integrations/Laminas/LaminasIntegration.php [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+                            $span->meta[Tag::COMPONENT] = 'laminas';
+                        }
+                    );
+                }
+
+                $rootSpan->resource = "$controller@$action $routeName";
+
+                $rootSpan->meta[Tag::HTTP_METHOD] = $method;
+                $rootSpan->meta[Tag::HTTP_URL] = Normalizer::urlSanitize($request->getUriString());
+                $rootSpan->meta['laminas.route.name'] = $routeName;
```
Comment (pierotibou) [inline src/Integrations/Integrations/Laminas/LaminasIntegration.php:207]: Check the span attributes tag doc I shared with you, I suppose those tags should be `http.` not `laminas.`

File: src/Integrations/Integrations/Laminas/LaminasIntegration.php [hunk 6]
```diff
@@ -180,10 +187,18 @@ function (SpanData $span, $args, $retval) use ($rootSpan) {
...
 
-                $method = $request->getMethod();
 
-                /** @var RouteMatch $routeMatch */
+                $rootSpan->meta[Tag::HTTP_METHOD] = $request->getMethod();
+                $rootSpan->meta[Tag::HTTP_VERSION] = $request->getVersion();
+                $rootSpan->meta[Tag::HTTP_URL] = Normalizer::urlSanitize($request->getUriString());
+
                 $routeMatch = $retval;
+                if (is_null($routeMatch)) {
```
Comment (pierotibou) [inline src/Integrations/Integrations/Laminas/LaminasIntegration.php:196]: You're not setting the resource name in that case which I believe is an issue

Comment (pierotibou) [reply src/Integrations/Integrations/Laminas/LaminasIntegration.php:196]: On seen together, it's not a pb

**Review by pierotibou — APPROVED** [review]

**Pull request merged**